cmd/gcp-controller-manager/app/gke_approver.go

```diff
@@ -370,5 +396,188 @@ func loadApproverOptions(s *GCPControllerManager) (approverOptions, error) {
+		if _, ok := blocks[name]; !ok {
+			return false
+		}
+	}
+	return true
+}
+
+func validateTPMAttestation(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	blocks, err := parsePEMBlocks(csr.Spec.Request)
+	if err != nil {
+		glog.Errorf("parsing csr.Spec.Request: %v", err)
+		return false
+	}
+	nodeIDRaw := blocks["VM IDENTITY"].Bytes
+	attestDataRaw := blocks["ATTESTATION DATA"].Bytes
+	attestSig := blocks["ATTESTATION SIGNATURE"].Bytes
+	attestCert := blocks["ATTESTATION CERTIFICATE"].Bytes
+
+	var nodeID node.Identity
+	if err := json.Unmarshal(nodeIDRaw, &nodeID); err != nil {
+		glog.Errorf("unmarshaling VM identity JSON: %v", err)
+		return false
+	}
+	if fmt.Sprint(nodeID.ProjectName) != opts.projectID {
+		glog.Errorf("received CSR for a different project Name (%d)", nodeID.ProjectName)
+		return false
+	}
+
+	client := oauth2.NewClient(context.Background(), opts.tokenSource)
+	cs, err := compute.New(client)
+	if err != nil {
+		glog.Errorf("creating GCE API client: %v", err)
+		return false
+	}
+	srv := compute.NewInstancesService(cs)
+	inst, err := srv.Get(fmt.Sprint(nodeID.ProjectID), nodeID.Region, nodeID.Name).Do()
+	if err != nil {
+		glog.Errorf("fetching VM data from GCE API: %v", err)
+		return false
+	}
+	if validateClusterMembership {
+		ok, err := clusterHasInstance(client, opts.projectID, inst.Zone, opts.clusterName, inst.Name)
+		if err != nil {
+			glog.Errorf("checking VM membership in cluster: %v", err)
+			return false
+		}
+		if !ok {
+			glog.Errorf("VM %q doesn't belong to cluster %q", inst.Name, opts.clusterName)
+			return false
+		}
+	}
+
+	// TODO(awly): get EK public key from GCE API.
+	// TODO(awly): verify ekCert to root CA.
+	ekCert, err := x509.ParseCertificate(attestCert)
+	if err != nil {
+		glog.Errorf("parsing ATTESTATION_CERTIFICATE: %v", err)
+		return false
+	}
+	ekPub, ok := ekCert.PublicKey.(*rsa.PublicKey)
+	if !ok {
+		glog.Errorf("public key in ATTESTATION CERTIFICATE is %T, want *rsa.PublicKey", ekCert.PublicKey)
+		return false
+	}
+
+	attestHash := sha256.Sum256(attestDataRaw)
+	if err := rsa.VerifyPKCS1v15(ekPub, crypto.SHA256, attestHash[:], attestSig); err != nil {
+		glog.Errorf("verifying certification signature with EK public key: %v", err)
+		return false
+	}
+
+	// Verify that attestDataRaw matches certificate.
+	attestData, err := tpm2.DecodeAttestationData(attestDataRaw)
+	if err != nil {
+		glog.Errorf("parsing attestation data in CSR: %v", err)
+		return false
+	}
+	pub, err := tpmattest.MakePublic(x509cr.PublicKey)
+	if err != nil {
+		glog.Errorf("converting public key in CSR to TPM Public structure: %v", err)
+		return false
+	}
+	ok, err = attestData.AttestedCertifyInfo.Name.MatchesPublic(pub)
+	if err != nil {
+		glog.Errorf("comparing subject name in attestation data to CSR public key: %v", err)
+		return false
+	}
+	if !ok {
+		glog.Errorf("subject name in attestation data doesn't match CSR public key")
```

Comment:
It's strange that the digest of a certified public key is called "subject name". Is that common?